[improve][client] Add Builder Methods to Create Message-based TableView

namest504 opened the pull request:
<!--
### Contribution Checklist
  
  - PR title format should be *[type][component] summary*. For details, see *[Guideline - Pulsar PR Naming Convention](https://pulsar.apache.org/contribute/develop-semantic-title/)*. 

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.
-->

<!-- Either this PR fixes an issue, -->

Fixes: #24744

<!-- or this PR is one task of an issue -->

Main Issue: #24744

<!-- If the PR belongs to a PIP, please add the PIP link here -->

PIP: 445 #24842

<!-- Details of when a PIP is required and how the PIP process work, please see: https://github.com/apache/pulsar/blob/master/pip/README.md -->

### Motivation

The current `TableView` API only exposes the deserialized message value, which limits access to essential message metadata like properties, event time, or the raw message object itself.

This PR introduces a flexible, non-breaking mechanism to create value-mapped views over a topic. It replaces the initial proposal of adding a `getRawMessage()` method, which would have performance implications for all users. Instead, it adds a `createMapped` method to the `TableViewBuilder`, allowing users to transform a full `Message<T>` into any custom object `V` that suits their needs. This provides maximum flexibility, from accessing the full raw message to creating custom, memory-efficient objects.

<!-- Explain here the context, and why you're making that change. What is the problem you're trying to solve. -->

### Modifications

- **Added `createMapped` and `createMappedAsync` to `TableViewBuilder`**: These methods accept a `Function<Message<T>, V>` mapper, letting users define their own mapping from a message to the value stored in the `TableView`. A `null` mapper is rejected with `IllegalArgumentException` (failed future in the async variant), and a `null` return value from the mapper is treated as a tombstone.
- **Moved the core logic to `AbstractTableViewImpl<T, V>`**: `TableViewImpl` keeps its name and becomes a thin subclass, so the behavior for existing users (including message pooling and release) is unchanged. The class naming follows the structure from @lhotari's experiment branch.
- **Introduced `MessageMapperTableViewImpl<T, V>`**: Implements the mapped views. Message pooling is disabled for mapped views since the mapper may retain the `Message` instance (e.g. with `Function.identity()`), so the pooled payload buffer must not be reused or released.

<!-- Describe the modifications you've done. -->

### Verifying this change

- Unit tests: `TableViewBuilderImplTest.testCreateMapped*` (mapped view creation, null mapper validation for both sync and async variants).
- Integration tests in `TableViewTest`: `testCreateMapped` (custom mapping, value updates, tombstone via `null` mapping) and `testCreateMappedWithIdentityMapper` (`TableView<Message<T>>` with metadata access).
- Verified locally with `./gradlew quickCheck` and the test classes above.

- [ ] Make sure that the change passes the CI checks.
<!--
*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*
-->

### Does this pull request potentially affect one of the following parts:

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [x] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. -->
- [x] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [ ] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

### Matching PR in forked repository

PR in forked repository: [fork-repo](https://github.com/namest504/pulsar)

<!--
After opening this PR, the build in apache/pulsar will fail and instructions will
be provided for opening a PR in the PR author's forked repository.

apache/pulsar pull requests should be first tested in your own fork since the 
apache/pulsar CI based on GitHub Actions has constrained resources and quota.
GitHub Actions provides separate quota for pull requests that are executed in 
a forked repository.

The tests will be run in the forked repository until all PR review comments have
been handled, the tests pass and the PR is approved by a reviewer.
-->

namest504 committed:
[improve][api] Add getRawMessage() method to TableView for accessing raw Pulsar message

namest504 committed:
[improve][impl] Fix TableViewImpl.handleMessage message lifecycle management by adding retain() before storing message to rawMessages

- Prevent premature message release that causes getRawMessage to return null keys.
- Ensure message reference count is balanced by retaining message before storing and releasing old message after replacement.
- Keep message release in finally block for safety.

lhotari reviewed (changes requested):
Thanks for the contribution!

Since this changes the Pulsar API, we'd need to first go through the [PIP process](https://github.com/apache/pulsar/tree/master/pip#pulsar-improvement-proposal-pip) to modify the public API.

Instead of adding a separate ConcurrentMap for storing the messages, I'd suggest going with an approach where the TableViewBuilder would have separate methods for creating a view for messages. 

```java
TableView<Message<T>> createForMessages() throws PulsarClientException;
CompletableFuture<TableView<Message<T>>> createForMessagesAsync() throws PulsarClientException;
```

I'm not exactly sure if there's any obstacles with this approach, but it seems that it could be a better way forward so that the existing runtime behavior of TableView implementation wouldn't change (like it does with the current PR changes).

lhotari commented:
> Hi @lhotari,
> 
> Thanks for the review and the great suggestion about using the `TableViewBuilder`!
> 
> I completely agree that we should avoid impacting the runtime behavior for existing users.
> 
> As you recommended, I've created a PIP for this API change, which you can find here: #24842
> 
> Let's continue the design discussion over on the PIP PR.

@namest504 you seemed to ignore the suggestion in the comment. I'd suggest revisiting the PIP accordingly and renaming it.

namest504 committed:
Refactor TableView to use builder for Message access

namest504 committed:
[improve][api] Add Builder Methods to Create Message-based TableView

Signed-off-by: namest504 <namest504@gmail.com>

namest504 committed:
[improve][api] Add Test

Signed-off-by: namest504 <namest504@gmail.com>

namest504 committed:
[improve][api] Rename AbstractTableView

Signed-off-by: namest504 <namest504@gmail.com>

namest504 committed:
[improve][api] Remove createForMessages method for using createMapped method

Signed-off-by: namest504 <namest504@gmail.com>

namest504 committed:
Disable message pooling for mapped table views and validate the mapper argument

- The mapper may retain the Message instance (e.g. Function.identity()), so the
  reader must not use pooled messages for mapped views; the payload buffer would
  otherwise be released or leaked. The classic TableView keeps using pooled
  messages and releasing them as before.
- Reject a null mapper with IllegalArgumentException (failed future in the async
  variant) instead of failing later in the async pipeline.
- Add unit tests for createMapped/createMappedAsync in TableViewBuilderImplTest.

namest504 committed:
Merge branch 'master' of https://github.com/apache/pulsar into master

Resolve conflicts by rebuilding the PIP-445 changes on top of current master:
- Keep the original TableViewImpl class name; move the shared logic to AbstractTableViewImpl
- Rename MessageMapperTableView to MessageMapperTableViewImpl
- Drop the unused MessageTableView class and the import/FQN churn

namest504 commented on pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TableViewTest.java:
```diff
@@ -53,7 +53,6 @@
 import org.apache.pulsar.client.api.PulsarClientException;
 import org.apache.pulsar.client.api.Reader;
 import org.apache.pulsar.client.api.Schema;
-import org.apache.pulsar.client.api.TableView;
```
Fixed. The import removal was caused by naming the impl class `TableView`, which shadowed the API interface in the same package. I went back to the `TableViewImpl` / `AbstractTableViewImpl` / `MessageMapperTableViewImpl` naming from your experiment branch, so the imports and the related diff churn are gone. The remaining changes in this file are the reflection target (`readAllExistingMessages` now lives in `AbstractTableViewImpl`) and the two new tests.

namest504 commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewBuilderImpl.java:
```diff
@@ -22,14 +22,10 @@
-import org.apache.pulsar.client.api.CryptoKeyReader;
-import org.apache.pulsar.client.api.PulsarClientException;
-import org.apache.pulsar.client.api.Schema;
-import org.apache.pulsar.client.api.TableView;
```
Fixed, same root cause as above. The explicit import list is restored and the wildcard import is gone.

namest504 commented on pulsar-client/src/test/java/org/apache/pulsar/client/impl/TableViewImplTest.java:
```diff
@@ -24,7 +24,6 @@
 import static org.testng.Assert.assertNotNull;
 import org.apache.pulsar.client.api.CryptoKeyReader;
 import org.apache.pulsar.client.api.Schema;
-import org.apache.pulsar.client.api.TableView;
```
Fixed. This file now has no changes compared to master.

namest504 commented:
@lhotari I've addressed the review comments and resolved the conflicts with master (rebuilt the refactoring on top of the current Gradle/slog codebase).

Main changes since your last review:
- Kept the original `TableViewImpl` class name and moved the shared logic to a package-private `AbstractTableViewImpl<T, V>`, following the naming in your experiment branch. No imports are removed and there are no fully qualified class name references anymore.
- Removed the unused `MessageTableView` class left over from the earlier `createForMessages` approach.
- Disabled message pooling for mapped views: the mapper may retain the `Message` instance, so the pooled payload buffer must not be reused or released. The classic `TableView` path keeps pooling and releasing as before.
- Added mapper null-validation and unit tests in `TableViewBuilderImplTest`.

The vote result for PIP-445 has been posted on the dev mailing list (3 binding +1s), and I've also updated the PIP document in #24842 to match the final class names. PTAL when you have a chance.

namest504 committed:
Fix reflection field reads in tests after moving fields to AbstractTableViewImpl

FieldUtils.readDeclaredField only looks at the exact class, so reading the
reader/compactionStrategy fields from TableViewImpl fails now that they live
in AbstractTableViewImpl. Use readField, which traverses the class hierarchy.
